torch/csrc/autograd/function.h

```diff
@@ -23,164 +17,360 @@
+/// functions in PyTorch's autograd machinery derive from this class and
+/// override its `apply` method. Instances of such subclasses will then be
+/// invokeable via the call operator.
+///
+///                    Functions in the Autograd Graph
+///~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// When viewing the autograd system as a graph, `Function`s are the vertices or
+/// nodes, connected to each other via (directed) `Edge`s, which themselves are
+/// represented via (`Function`, input_nr) pairs. `Variable`s are the outputs to
+/// and inputs of `Function`s, and travel between these edges during execution
+/// of the graph. When two or more `Edge`s (from different sources) point at the
+/// same input to a `Function`, the values produced along all of these edges are
+/// implicitly summed prior to being forwarded to the target `Function`.
+///
+///                              Hierarchy
+///~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// Subclasses will often represent forward passes (functions) or backward
+/// passes (derivatives). Note, however, that due to the very general
+/// definition of a `Function` taking *zero* or more inputs and producing
+/// *zero* or more outputs, uses of `Function`s are flexible and extend beyond
+/// purely mathematical operations. For example, the `AccumulateGrad` function
+/// is a *sink*: it takes one input, but produces no outputs, instead adding
+/// the input to its internal accumulator as a side effect. At the other
+/// extreme, the `GraphRoot` function receives no inputs from other functions,
+/// but produces multiple outputs.
+///
+///                              Interface
+///~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// The most important method on `Function` is the call operator, which takes in
+/// a list of variables and produces a list of variables. The precise size of
+/// these lists can be determined with `num_inputs()` and `num_outputs()`.
+/// `Function`s are stitched together via their `next_edge` interface, which let
+/// you manipulate the set of outgoing edges of a `Function`. You can add an
+/// edge with `add_next_edge()`, retrieve an edge with `next_edge(index)` and
+/// iterate over them via the `next_edges()` method. Other methods exist for
+/// integration with the JIT and other parts of PyTorch. Every `Function` has a
+/// *sequence number* that increases monotonically in the order of `Function`
+/// construction. It can be retrieved via the `sequence_nr()` method. Note that
+/// this sequence number is *thread local*. This means that when `Function`s
+/// `A`, `B` and `C` are created consecutively in the same thread, their
+/// sequence numbers will be ordered `A` < `B` < `C`. If, however, `A` and `B`
+/// are created in one thread and `C` is created in a new thread, `C` will have
+/// a *lower* sequence number than `B`.
```

Comment:
I think it's just better to say that we have no guarantees wrt consistency of sequence numbers between threads. It's not *guaranteed to be lower*. It just happens to be so.